File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.py
from heapq import heappush, heappop


def dijkstra(graph, n, start, end):
    heap = []
    
    INF = 2 * int(1e7) + 1
    costs = [INF] * (n + 1)
    costs[start] = 0
    
    heappush(heap, (0, start))
    
    while heap:
        cur_cost, sour = heappop(heap)
        
        if cur_cost > costs[sour]:
            continue
        
        for des, cost in graph[sour]:
            next_cost = cost + cur_cost
            
            if next_cost > costs[des]:
                continue
                
            costs[des] = next_cost
            heappush(heap, (next_cost, des))
    
    return costs[end]

def solution(n, s, a, b, fares):
    answer = 2 * int(1e7)
    graph = [[] for _ in range(n + 1)]
    
    for sour, des, cost in fares:
        graph[sour].append((des, cost))
        graph[des].append((sour, cost))
           
    for mid in range(1, n + 1):
        cost1 = dijkstra(graph, n, s, mid)
        cost2 = dijkstra(graph, n, mid, a)
        cost3 = dijkstra(graph, n, mid, b)
    
        answer = min(answer, cost1 + cost2 + cost3)
    
    return answer

File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.java
import java.util.*;

class Solution {
    List<int[]>[] graph;
    public int solution(int n, int s, int a, int b, int[][] fares) {
        int answer = 0;
        int[][] costs = new int[n + 1][n + 1];
        graph = new ArrayList[n + 1];
        
        for(int i=1 ; i<=n ; i++){
            graph[i] = new ArrayList<>();
            Arrays.fill(costs[i], Integer.MAX_VALUE);
            costs[i][i] = 0;
        }
        
        for(int[] fare : fares){
            int x = fare[0], y = fare[1];    
            costs[x][y] = costs[y][x] = fare[2];
            graph[x].add(new int[]{y, fare[2]});
            graph[y].add(new int[]{x, fare[2]});
        }
        
        for(int k=1 ; k<=n ; k++){
            for(int i=1 ; i<=n ; i++){
                for(int j=1 ; j<=n ; j++){
                    if(costs[i][k] == Integer.MAX_VALUE 
                       || costs[k][j] == Integer.MAX_VALUE)
                        continue;
                    
                    costs[i][j] = Math.min(costs[i][j], costs[i][k] + costs[k][j]);
                }
            }
        }
    
        return dijkstra(s, a, b, costs);
    }
    
    int dijkstra(int start, int a, int b, int[][] costs){
        Queue<int[]> que = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]);
        que.add(new int[]{start, 0});
        
        int[] dists = new int[costs.length];
        Arrays.fill(dists, Integer.MAX_VALUE);
        dists[start] = 0;
        
        int ret = costs[start][a] + costs[start][b];
        while(!que.isEmpty()){
            int[] cur = que.poll();
            
            int sour = cur[0], dist = cur[1];
            if(dist > dists[sour])
                continue;
            
            for(int[] next : graph[sour]){
                int des = next[0];
                
                if(dists[des] <= dist + next[1])
                    continue;
                
                ret = Math.min(ret, dist + next[1] + costs[des][a] + costs[des][b]);
                dists[des] = dist + next[1];
                que.add(new int[]{des, dist + next[1]});
            }
        }
        
        return ret;
    }
}